politeiawww/pi/error.go

```diff
@@ -0,0 +1,197 @@
+			v1.UserErrorReply{
+				ErrorCode:    ue.ErrorCode,
+				ErrorContext: ue.ErrorContext,
+			})
+		return
+
+	case errors.As(err, &pe):
+		// politeiawww plugin error
+		m := fmt.Sprintf("%v Plugin error: %v %v",
+			util.RemoteAddr(r), pe.PluginID, pe.ErrorCode)
+		if pe.ErrorContext != "" {
+			m += fmt.Sprintf(": %v", pe.ErrorContext)
+		}
+		log.Infof(m)
+		util.RespondWithJSON(w, http.StatusBadRequest,
+			v1.PluginErrorReply{
+				PluginID:     pe.PluginID,
+				ErrorCode:    pe.ErrorCode,
+				ErrorContext: pe.ErrorContext,
+			})
+		return
+
+	case errors.As(err, &pde):
+		// Politeiad error
+		handlePDError(w, r, format, pde)
+
+	default:
+		// Internal server error. Log it and return a 500.
+		t := time.Now().Unix()
+		e := fmt.Sprintf(format, err)
+		log.Errorf("%v %v %v %v Internal error %v: %v",
+			util.RemoteAddr(r), r.Method, r.URL, r.Proto, t, e)
+
+		// If this is a pkg/errors error then we can pull the
+		// stack trace out of the error, otherwise, we use the
+		// stack trace for this function.
+		stack, ok := util.StackTrace(err)
+		if !ok {
+			stack = string(debug.Stack())
+		}
+
+		log.Errorf("Stacktrace (NOT A REAL CRASH): %v", stack)
+
+		util.RespondWithJSON(w, http.StatusInternalServerError,
+			v1.ServerErrorReply{
+				ErrorCode: t,
+			})
+		return
+	}
+}
+
+func handlePDError(w http.ResponseWriter, r *http.Request, format string, pde pdclient.RespError) {
+	var (
+		pluginID   = pde.ErrorReply.PluginID
+		errCode    = pde.ErrorReply.ErrorCode
+		errContext = pde.ErrorReply.ErrorContext
+	)
+	e := convertPDErrorCode(errCode)
+	switch {
+	case pluginID != "":
+		// politeiad plugin error. Log it and return a 400.
+		m := fmt.Sprintf("%v Plugin error: %v %v",
+			util.RemoteAddr(r), pluginID, errCode)
+		if errContext != "" {
+			m += fmt.Sprintf(": %v", errContext)
+		}
+		log.Infof(m)
+		util.RespondWithJSON(w, http.StatusBadRequest,
+			v1.PluginErrorReply{
+				PluginID:     pluginID,
+				ErrorCode:    errCode,
+				ErrorContext: errContext,
+			})
+		return
+
+	case e != v1.ErrorCodeInvalid:
+		// User error from politeiad that corresponds to a records user
+		// error. Log it and return a 400.
+		m := fmt.Sprintf("%v Records user error: %v %v",
+			util.RemoteAddr(r), e, v1.ErrorCodes[e])
+		if errContext != "" {
+			m += fmt.Sprintf(": %v", errContext)
+		}
+		log.Infof(m)
+		util.RespondWithJSON(w, http.StatusBadRequest,
+			v1.UserErrorReply{
+				ErrorCode:    e,
+				ErrorContext: errContext,
+			})
+		return
+
+	default:
+		// politeiad error does not correspond to a user error. Log it
+		// and return a 500.
+		ts := time.Now().Unix()
+		log.Errorf("%v %v %v %v Internal error %v: error code "+
+			"from politeiad: %v", util.RemoteAddr(r), r.Method, r.URL,
+			r.Proto, ts, errCode)
+
+		util.RespondWithJSON(w, http.StatusInternalServerError,
+			v1.ServerErrorReply{
+				ErrorCode: ts,
+			})
+		return
+	}
+}
+
+func convertPDErrorCode(errCode uint32) v1.ErrorCodeT {
```

Comment:
maps PD errors ?

Reply:
You only need to map the errors that could be returned from the politeiad plugin cmd routes.